File: src/Conjecture.FSharp.Tests/ReproExportIntegrationTests.fs
// Copyright (c) 2026 Kim Ommundsen. Licensed under the MPL-2.0.
// See LICENSE.txt in the project root or https://mozilla.org/MPL/2.0/

module Conjecture.FSharp.Tests.ReproExportIntegrationTests

open System
open System.IO
open Xunit
open Conjecture
open Conjecture.Core

[<Fact>]
let ``runBoolWithRepro failing property writes a .cs repro file when ExportOnFailure is true`` () =
    task {
        let tempDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName())
        Directory.CreateDirectory(tempDir) |> ignore
        try
            let repro = { ExportOnFailure = true; OutputPath = tempDir }
            let genWrapped = Gen.int (0, 100)
            let! result = PropertyRunner.runBoolWithRepro genWrapped (fun n -> n < 5) repro
            match result with
            | PropertyResult.Failed _ ->
                let files = Directory.GetFiles(tempDir, "*.cs")
                Assert.NotEmpty(files)
            | PropertyResult.Passed ->
                Assert.Fail("Expected property to fail")
        finally
            try Directory.Delete(tempDir, true) with _ -> ()
    }

[<Fact>]
let ``runBoolWithRepro failing property writes no file when ExportOnFailure is false`` () =
    task {
        let tempDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName())
        Directory.CreateDirectory(tempDir) |> ignore
        try
            let repro = { ExportOnFailure = false; OutputPath = tempDir }
            let genWrapped = Gen.int (0, 100)
            let! result = PropertyRunner.runBoolWithRepro genWrapped (fun n -> n < 5) repro
            match result with
            | PropertyResult.Failed _ ->
                let files = Directory.GetFiles(tempDir, "*.cs")
                Assert.Empty(files)
            | PropertyResult.Passed ->
                Assert.Fail("Expected property to fail")
        finally
            try Directory.Delete(tempDir, true) with _ -> ()
    }

[<Fact>]
let ``runUnitWithRepro failing property writes a .cs repro file when ExportOnFailure is true`` () =
    task {
        let tempDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName())
        Directory.CreateDirectory(tempDir) |> ignore
        try
            let repro = { ExportOnFailure = true; OutputPath = tempDir }
            let genWrapped = Gen.int (0, 100)
            let! result =
                PropertyRunner.runUnitWithRepro genWrapped (fun n ->
                    if n >= 5 then raise (Exception("fail"))) repro
            match result with
            | PropertyResult.Failed _ ->
                let files = Directory.GetFiles(tempDir, "*.cs")
                Assert.NotEmpty(files)
            | PropertyResult.Passed ->
                Assert.Fail("Expected property to fail")
        finally
            try Directory.Delete(tempDir, true) with _ -> ()
    }

[<Fact>]
let ``ReproExport disabled defaults are sensible`` () =
    Assert.False(ReproExport.disabled.ExportOnFailure)
    Assert.Equal(".conjecture/repros/", ReproExport.disabled.OutputPath)

File: src/Conjecture.Xunit.V3.Tests/ReproExportIntegrationTests.cs
// Copyright (c) 2026 Kim Ommundsen. Licensed under the MPL-2.0.
// See LICENSE.txt in the project root or https://mozilla.org/MPL/2.0/

using System.Reflection;

using Conjecture.Core;
using Conjecture.Core.Internal;

using Xunit;

namespace Conjecture.Xunit.V3.Tests;

/// <summary>
/// Integration tests verifying the xUnit v3 adapter exposes the repro-export
/// surface and that the underlying ReproFileBuilder + TestRunner pipeline writes
/// a .cs file when the property fails and ExportReproductionOnFailure is true.
/// </summary>
public class ReproExportIntegrationTests
{
#pragma warning disable IDE0060
    private static void IntProperty(int x) { }
#pragma warning restore IDE0060

    private static ParameterInfo[] Params(string methodName)
    {
        return typeof(ReproExportIntegrationTests)
            .GetMethod(methodName, BindingFlags.NonPublic | BindingFlags.Static)!
            .GetParameters();
    }

    [Fact]
    public void PropertyAttribute_ExportReproductionOnFailure_DefaultIsFalse()
    {
        PropertyAttribute attr = new();
        Assert.False(attr.ExportReproductionOnFailure);
    }

    [Fact]
    public void PropertyAttribute_ReproductionOutputPath_DefaultIsConjecturePath()
    {
        PropertyAttribute attr = new();
        Assert.Equal(".conjecture/repros/", attr.ReproductionOutputPath);
    }

    [Fact]
    public async Task ExportReproductionOnFailureTrue_FailingProperty_WritesCsFile()
    {
        string tempDir = Path.Combine(Path.GetTempPath(), Path.GetRandomFileName());
        Directory.CreateDirectory(tempDir);
        try
        {
            ConjectureSettings settings = new()
            {
                MaxExamples = 50,
                Seed = 1UL,
                ExportReproductionOnFailure = true,
                ReproductionOutputPath = tempDir,
            };
            ParameterInfo[] parameters = Params(nameof(IntProperty));

            TestRunResult result = await TestRunner.Run(settings, data =>
            {
                object[] args = SharedParameterStrategyResolver.Resolve(parameters, data);
                if ((int)args[0] > 5) { throw new Exception("fail"); }
            });

            Assert.False(result.Passed);

            ConjectureData replay = ConjectureData.ForRecord(result.Counterexample!);
            object[] shrunkArgs = SharedParameterStrategyResolver.Resolve(parameters, replay);
            IEnumerable<(string Name, object? Value, Type Type)> paramTuples =
                parameters.Select((p, i) => (p.Name!, (object?)shrunkArgs[i], p.ParameterType));
            ReproContext context = new(
                nameof(ReproExportIntegrationTests),
                nameof(IntProperty),
                false,
                paramTuples,
                result.Seed!.Value,
                result.ExampleCount,
                result.ShrinkCount,
                TestFramework.Xunit,
                DateTimeOffset.UtcNow);

            ReproFileBuilder.WriteToFile(context, settings.ReproductionOutputPath);
            string[] files = Directory.GetFiles(tempDir, "*.cs");
            Assert.NotEmpty(files);
        }
        finally
        {
            Directory.Delete(tempDir, true);
        }
    }
}
